libs/langchain-mongodb/langchain_mongodb/graphrag/graph.py

```diff
@@ -0,0 +1,482 @@
+        chain: RunnableSequence = self.entity_prompt | self.entity_extraction_model
+        # Invoke on a document to extract entities and relationships
+        response: AIMessage = chain.invoke(
+            dict(
+                input_document=raw_document,
+                entity_schema=self.entity_schema,
+                allowed_entity_types=self.allowed_entity_types,
+                allowed_relationship_types=self.allowed_relationship_types,
+            )
+        )
+        # Post-Process output string into list of entity json documents
+        # Strip the ```json prefix and trailing ```
+        json_string = (
+            response.content.removeprefix("```json").removesuffix("```").strip()
+        )
+        extracted = json.loads(json_string)
+        return extracted["entities"]
+
+    def extract_entity_names(self, raw_document: str, **kwargs: Any) -> List[str]:
+        """Extract entity names from a document for similarity_search.
+
+        The second entity extraction has a different form and purpose than
+        the first as we are looking for starting points of our search and
+        paths to follow. We aim to find source nodes,  but no target nodes or edges.
+
+        Args:
+            raw_document: A single text document as a string. Typically prose.
+        Returns:
+            List of entity names / _ids.
+        """
+        # Combine the llm with the prompt template to form a chain
+        chain: RunnableSequence = self.query_prompt | self.entity_extraction_model
+        # Invoke on a document to extract entities and relationships
+        response: AIMessage = chain.invoke(
+            dict(
+                input_document=raw_document,
+                allowed_entity_types=self.allowed_entity_types,
+            )
+        )
+        # Post-Process output string into list of entity json documents
+        # Strip the ```json prefix and suffix
+        json_string = (
+            response.content.removeprefix("```json").removesuffix("```").strip()
+        )
+        return json.loads(json_string)
+
+    def find_entity_by_name(self, name: str) -> Optional[Entity]:
+        """Utility to get Entity dict from Knowledge Graph / Collection.
+        Args:
+            name: _id string to look for
+        Returns:
+            List of Entity dicts if any match name
+        """
+        return self.collection.find_one({"_id": name})
+
+    def related_entities(
+        self,
+        starting_entities: List[str],
+        max_depth: Optional[int] = None,
+    ) -> List[Entity]:
+        """Traverse Graph along relationship edges to find connected entities.
+
+        Args:
+            starting_entities: Traversal begins with documents whose _id fields match these strings.
+            max_depth: Recursion continues until no more matching documents are found,
+            or until the operation reaches a recursion depth specified by the maxDepth parameter
```

Comment:
does this need to be indented to render properly?

Reply:
I mean the line that continues after `max_depth`, I think it needs to be indented further